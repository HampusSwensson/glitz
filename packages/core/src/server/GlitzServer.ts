import { Style } from '@glitz/type';
import Base from '../core/Base';
import { DEFAULT_HYDRATION_IDENTIFIER, Options } from '../types/options';
import { createHashCounter } from '../utils/hash';
import InjectorServer from './InjectorServer';

export default class GlitzServer<TStyle = Style> extends Base<TStyle> {
  public getStyleMarkup: (className?: string) => string;
  constructor(options: Options = {}) {
    const prefix = options.prefix;
    const classHasher = createHashCounter(prefix);
    const keyframesHasher = createHashCounter(prefix);
    const fontFaceHasher = createHashCounter(prefix);

    let plain: InjectorServer;
    const mediaIndex: {
      [media: string]: InjectorServer;
    } = {};

    const injector = (media?: string) =>
      media
        ? (mediaIndex[media] = mediaIndex[media] || new InjectorServer(classHasher, keyframesHasher, fontFaceHasher))
        : (plain = plain || new InjectorServer(classHasher, keyframesHasher, fontFaceHasher));

    super(injector, options.transformer, options.atomic);

    this.getStyleMarkup = () => {
      const attribute = options.identifier || DEFAULT_HYDRATION_IDENTIFIER;
      let markup = '';
      if (plain) {
        markup += `<style data-${attribute}>${plain.getStyle()}</style>`;
      }
      const medias = options.mediaOrder ? Object.keys(mediaIndex).sort(options.mediaOrder) : Object.keys(mediaIndex);
      for (const media of medias) {
        markup += `<style data-${attribute} media="${media}">${mediaIndex[media].getStyle()}</style>`;
      }
      return markup;
    };
  }
}
