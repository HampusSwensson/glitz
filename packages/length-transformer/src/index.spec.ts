import GlitzServer from '@glitz/core/server';
import numberToLength, { createNumberToLengthTransformer } from './';

describe('number to unit', () => {
  it('injects atomic style with numeric values as px', () => {
    const server = new GlitzServer({ transformer: numberToLength });

    expect(
      server.injectStyle({
        width: 1,
        height: [1, 'max-content'],
        padding: { left: 1 },
        columns: 1,
        lineHeight: 1,
        left: 0,
      }),
    ).toBe('a b c d e f');
    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('injects atomic style with numeric values as rem', () => {
    const server = new GlitzServer({
      transformer: createNumberToLengthTransformer({ defaultUnit: 'rem', columns: 'em', paddingLeft: 'px' }),
    });

    expect(
      server.injectStyle({
        width: 1,
        height: [1, 'max-content'],
        padding: { left: 1 },
        columns: 1,
        lineHeight: 1,
        left: 0,
      }),
    ).toBe('a b c d e f');
    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('injects non-atomic style with numeric values as px', () => {
    const server = new GlitzServer({ transformer: numberToLength, atomic: false });

    expect(
      server.injectStyle({
        width: 1,
        height: [1, 'max-content'],
        padding: { left: 1 },
        columns: 1,
        lineHeight: 1,
        left: 0,
      }),
    ).toBe('a');
    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('injects non-atomic style with numeric values as rem', () => {
    const server = new GlitzServer({
      transformer: createNumberToLengthTransformer({ defaultUnit: 'rem', columns: 'em', paddingLeft: 'px' }),
      atomic: false,
    });

    expect(
      server.injectStyle({
        width: 1,
        height: [1, 'max-content'],
        padding: { left: 1 },
        columns: 1,
        lineHeight: 1,
        left: 0,
      }),
    ).toBe('a');
    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
});
