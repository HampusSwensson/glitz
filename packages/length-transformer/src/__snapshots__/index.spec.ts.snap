// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`number to unit injects atomic style with numeric values as px 1`] = `"<style class=\\"__glitz__\\">.a{width:1px}.b{height:1px;height:max-content}.c{padding-left:1px}.d{columns:1}.e{line-height:1}.f{left:0}.g{order:1}</style>"`;

exports[`number to unit injects atomic style with numeric values as rem 1`] = `"<style class=\\"__glitz__\\">.a{width:1rem}.b{height:1rem;height:max-content}.c{padding-left:1px}.d{columns:1em}.e{line-height:1}.f{left:0}.g{order:1}</style>"`;

exports[`number to unit injects non-atomic style with numeric values as px 1`] = `"<style class=\\"__glitz__\\">.a{width:1px;height:1px;height:max-content;padding-left:1px;columns:1;line-height:1;left:0;order:1}</style>"`;

exports[`number to unit injects non-atomic style with numeric values as rem 1`] = `"<style class=\\"__glitz__\\">.a{width:1rem;height:1rem;height:max-content;padding-left:1px;columns:1em;line-height:1;left:0;order:1}</style>"`;
